src/util.h

```diff
@@ -178,77 +178,90 @@ inline TypeName* Unwrap(v8::Local<v8::Object> object);
 
 inline void SwapBytes(uint16_t* dst, const uint16_t* src, size_t buflen);
 
-class Utf8Value {
+// Allocates an array of member type T. For up to StackStorageSize items,
+// the stack is used, otherwise malloc().
+template <typename T, size_t kStackStorageSize>
+class MaybeStackBuffer {
   public:
-    explicit Utf8Value(v8::Isolate* isolate, v8::Local<v8::Value> value);
+    const T* out() const {
+      return buf_;
+    }
 
-    ~Utf8Value() {
-      if (str_ != str_st_)
-        free(str_);
+    T* out() {
+      return buf_;
     }
 
-    char* operator*() {
-      return str_;
-    };
+    // operator* for compatibility with `v8::String::(Utf8)Value`
+    T* operator*() {
+      return buf_;
+    }
 
-    const char* operator*() const {
-      return str_;
-    };
+    const T* operator*() const {
+      return buf_;
+    }
 
     size_t length() const {
       return length_;
-    };
-
-  private:
-    size_t length_;
-    char* str_;
-    char str_st_[1024];
-};
+    }
 
-class TwoByteValue {
-  public:
-    explicit TwoByteValue(v8::Isolate* isolate, v8::Local<v8::Value> value);
+    // Call to make sure enough space for `size` entries is available.
+    // There can only be 1 call to EnsureSufficientStorage or Invalidate
+    // per instance.
+    void EnsureSufficientStorage(size_t size) {
+      if (size <= kStackStorageSize) {
+        buf_ = buf_st_;
+      } else {
+        buf_ = static_cast<T*>(malloc(sizeof(T) * size));
+        CHECK_NE(buf_, nullptr);
+      }
+    }
 
-    ~TwoByteValue() {
-      if (str_ != str_st_)
-        free(str_);
+    void SetLength(size_t length) {
+      length_ = length;
     }
 
-    uint16_t* operator*() {
-      return str_;
-    };
+    void SetLengthAndZeroTerminate(size_t length) {
+      SetLength(length);
 
-    const uint16_t* operator*() const {
-      return str_;
-    };
+      // T() is 0 for integer types, nullptr for pointers, etc.
+      buf_[length] = T();
+    }
 
-    size_t length() const {
-      return length_;
-    };
+    // Make derefencing this object return nullptr.
+    // Calling this is mutually exclusive with calling EnsureSufficientStorage.
+    void Invalidate() {
+      length_ = 0;
+      buf_ = nullptr;
```

Comment:
I can add it if you want, but right now that would be a no-op. This is only used by `BufferValue`, and only when calling `EnsureSufficientStorage` does not make sense because the conversion failed completely.

I didn’t want to change the overall behaviour here too much, so I just added the comment above. If you say you’d rather have a check for `buf_ != buf_st_` added, then I’d like to add something similar in `EnsureSufficientStorage` for consistency.